ui/job-view/redux/stores/pushes.js

```diff
@@ -0,0 +1,385 @@
+      newPushList[newPushList.length - 1].push_timestamp;
+
+    const newStuff = {
+      pushList: newPushList,
+      oldestPushTimestamp,
+      ...doRecalculateUnclassifiedCounts(jobMap),
+      ...getRevisionTips(newPushList),
+    };
+
+    // since we fetched more pushes, we need to persist the push state in the URL.
+    const updatedLastRevision = newPushList[newPushList.length - 1].revision;
+
+    if (setFromchange && getUrlParam('fromchange') !== updatedLastRevision) {
+      const params = getAllUrlParams();
+      params.set('fromchange', updatedLastRevision);
+      replaceLocation(params);
+      // We are silently updating the url params, but we still want to
+      // update the ActiveFilters bar to this new change.
+      window.dispatchEvent(new CustomEvent(thEvents.filtersUpdated));
+    }
+
+    return newStuff;
+  }
+  return {};
+};
+
+const fetchNewJobs = () => {
+  return (dispatch, getState) => {
+    const {
+      pushes: { pushList, jobMap },
+    } = getState();
+
+    if (!pushList.length) {
+      // If we have no pushes, then no need to get jobs.
+      return;
+    }
+
+    const pushIds = pushList.map(push => push.id);
+    const lastModified = getLastModifiedJobTime(jobMap);
+
+    JobModel.getList(
+      {
+        push_id__in: pushIds.join(','),
+        last_modified__gt: lastModified.toISOString().replace('Z', ''),
+      },
+      { fetch_all: true },
+    ).then(data => {
+      const failureStatus = false;
+      if (!failureStatus) {
+        // break the jobs up per push
+        const jobs = data.reduce((acc, job) => {
+          const pushJobs = acc[job.push_id]
+            ? [...acc[job.push_id], job]
+            : [job];
+          return { ...acc, [job.push_id]: pushJobs };
+        }, {});
+        // If a job is selected, and one of the jobs we just fetched is the
+        // updated version of that selected job, then send that with the event.
+        const selectedJobId = getUrlParam('selectedJob');
+        const updatedSelectedJob = selectedJobId
+          ? data.find(job => job.id === parseInt(selectedJobId, 10))
+          : null;
+
+        window.dispatchEvent(
+          new CustomEvent(thEvents.applyNewJobs, {
+            detail: { jobs },
+          }),
+        );
+        if (updatedSelectedJob) {
+          dispatch(setSelectedJob(updatedSelectedJob));
+        }
+      }
+    });
+  };
+};
+
+const doUpdateJobMap = (jobList, jobMap, pushList) => {
+  if (jobList.length) {
+    // lodash ``keyBy`` is significantly faster than doing a ``reduce``
+    return {
+      jobMap: { ...jobMap, ...keyBy(jobList, 'id') },
+      jobsLoaded: pushList.every(push => push.jobsLoaded),
+    };
+  }
+  return {};
+};
+
+export const fetchPushes = (
+  count = DEFAULT_PUSH_COUNT,
+  setFromchange = false,
+) => {
+  return async (dispatch, getState) => {
+    const {
+      pushes: { pushList, jobMap, oldestPushTimestamp },
+    } = getState();
+
+    dispatch({ type: LOADING });
```

Comment:
Nit: why the ALL CAPS? :)

Reply:
Well, this is a convention that you make class-level or file-level constants all caps to give you a heads up when you're using something like that.  Here is how the AirBnB guide describes it:

https://github.com/airbnb/javascript/#naming--uppercase

I also like to use them, even if not exported, for things like numbers or keys that don't change.  I think it's more of a convention than a rule, really.  But I like it.  :)

Reply:
oh, interesting :)